rest/src/market_data_rest.py

```diff
@@ -0,0 +1,182 @@
+        self.public_path = public_path
+        self.private_path = private_path
+        # Optional, if doing only public queries then it is not required
+        self.public_key = public_key
+        self.private_key = private_key
+
+    @staticmethod
+    def _nonce():
+        """
+        Nonce counter.
+        :returns: int, an always-increasing unsigned integer (up to 64 bits wide)
+        """
+        return int(1000 * time.time())
+
+    def _query_public(self, method, timeout=10, headers=None, data=None, request_type=GET) -> json:
+        """
+        Used to query the public endpoints
+
+        :param method:
+        :param timeout:
+        :param headers:
+        :param data:
+        :param request_type: str, 'GET or 'POST'
+        :return:
+        """
+        if headers is None:
+            headers = {}
+        if data is None:
+            data = {}
+
+        # the url is composed of the base url + the route to public if any + the endpoint itself
+        url = self.api_url + self.public_path + method
+
+        response = make_request(self.session, url, timeout, headers, data, request_type)
+
+        if response.status_code == HTTPStatus.OK:
+            return response.json()
+        else:
+            return self.process_error(response)
+
+    def _query_private(self, method, timeout=10, headers=None, data=None, request_type=GET):
+        """
+        Query private information
+
+        :param method:
+        :param timeout:
+        :param headers:
+        :param data:
+        :param request_type:
+        :return:
+        """
+        if data is None:
+            data = {}
+        data['nonce'] = self._nonce()
+
+        method_path = self.private_path + method
+        url = self.api_url + method_path
+
+        if headers is None:
+            if not self.public_key:
+                raise Exception('Public key is not set!')
+            if not self.private_key:
+                raise Exception('Private key is not set!')
+            signature = self._sign(method_path)
+
+            headers = {
+                self.header_key_col: self.public_key,
+                self.header_signature_col: signature
+            }
+
+        response = make_request(self.session, url, timeout, headers, data, request_type)
+        if response.status_code == HTTPStatus.OK:
+            return response.json()
+        else:
+            return self.process_error(response)
+
+    @abstractmethod
+    def _sign(self, method_path):
+        """
+        Used to sign request to private endpoints. Each Exchange has its own method of signing
+
+        :param method_path: str
+        :return:
+        """
+        pass
+
+    @abstractmethod
+    def process_error(self, response):
+        """
+        Used to process an error received from the endpoint
+
+        :param response: session.response
+        :return:
+        """
+        pass
+
+    @abstractmethod
+    def get_tob_bid(self) -> float:
```

Comment:
From here down below is a list of functions we want for each market/instrument_type combination of APIs. What else can you think of @yvesleconte ?